File: back_end/fruitsapp_backend/Controllers/AuthController.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Security.Claims;
using System.Threading.Tasks;
using fruitsapp_backend.Models;
using fruitsapp_backend.Models.DTOs;
using fruitsapp_backend.Service.Interfaces;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using static System.Runtime.InteropServices.JavaScript.JSType;

// For more information on enabling Web API for empty projects, visit https://go.microsoft.com/fwlink/?LinkID=397860

namespace fruitsapp_backend.Controllers
{
    [Route("api/[controller]")]
    public class AuthController : Controller
    {
        private readonly IAuthService _authService;

        public AuthController(IAuthService authService)
        {
            _authService = authService;
        }


        // login
        [AllowAnonymous]
        [HttpPost]
        [Route("login")]
        public async Task<IActionResult> Login([FromBody] Login model)
        {
            try
            {
                var result =   await _authService.login(model);
                return Ok(new { status = "success", message = "Login successfully!", token = result });

            }
            catch (ArgumentException e)
            {
                // Bắt lỗi ArgumentException và trả về phản hồi BadRequest với thông điệp lỗi
                return BadRequest(new { status = "error", message = e.Message });
            }
            catch (Exception e)
            {
                return StatusCode(500, new { status = "error", message = "An error occurred while processing your request", error = e.Message});
            }
        }

        // resgiter
        [HttpPost]
        [Route("register-user")]
        public async Task<IActionResult> Register([FromForm] Register model)
        {
            try
            {
                var result =  await _authService.register_user(model);

                return Ok(new { status = "success", message = " Registered successfully", data = result });

            }
            catch (ArgumentException ex)
            {
                // Bắt lỗi ArgumentException và trả về phản hồi BadRequest với thông điệp lỗi
                return BadRequest(new { status = "error", message = ex.Message });
            }
            catch (Exception e)
            {
                return StatusCode(500, new { status = "error", message = "An error occurred while processing your request", error = e.Message, stackTrace = e.StackTrace });
            }
        }

        [HttpGet]
        [Route("get-user")]
        [Authorize(Policy = "AdminOrUser")] // Httpcontext.Request.Headers["Authorization"] to get token
        public  async Task<IActionResult> GetUser()
        {
            try
            {
                var userData = await GetCurrentUser();
                return Ok(new { status = "success", message = " Authenticated! successfully", user = userData });
            }
            catch(ArgumentException e)
            {
                return BadRequest(new { status = "error", message = e.Message });
            }
            catch (UnauthorizedAccessException)
            {
                return Unauthorized(new { status = "error", message = "Unauthorized access. Please check your credentials and try again." });
            }
            catch (Exception e)
            {
                return StatusCode(500, new { status = "error", message = "An error occurred while processing your request", error = e.Message, stackTrace = e.StackTrace });
            }
        }

        private async Task <User> GetCurrentUser()
        {
            var identity =  HttpContext.User.Identity as ClaimsIdentity;

            if (identity != null)
            {
                var useridClaim = identity.Claims.FirstOrDefault(x => x.Type == "UserId")!.Value;
                var emailClaim = identity.Claims.FirstOrDefault(x => x.Type == ClaimTypes.Email)!.Value;
                var roleidClaim = identity.Claims.FirstOrDefault(x => x.Type == "RoleId")!.Value;
                var roleClaim = identity.Claims.FirstOrDefault(x => x.Type == ClaimTypes.Role)!.Value;

                if (emailClaim != null && roleClaim != null)
                {
                    var user = await _authService.getUser(int.Parse(useridClaim), emailClaim, int.Parse(roleidClaim));
                   
                    return user;
                }
            }

            return null;
        }

        [AllowAnonymous]
        [HttpPost]
        [Route("user-logout")]
        public async Task<IActionResult> Logout([FromBody] string token)
        {
            try
            {
                var result = await _authService.Logout(token);

                return Ok(new { status = "success", message = "Logout successfully!"});

            }
            catch (ArgumentException e)
            {
                // Bắt lỗi ArgumentException và trả về phản hồi BadRequest với thông điệp lỗi
                return BadRequest(new { status = "error", message = e.Message });
            }
            catch (Exception e)
            {
                return StatusCode(500, new { status = "error", message = "An error occurred while processing your request", error = e.Message });
            }
        }

       

        [HttpPost]
        [Route("update-profile-auth")]

        public async Task<IActionResult> UpdateProfile([FromForm] UserProfile modelRequest)
        {
            try
            {
                
                var result = await _authService.update_profile(modelRequest);

                return Ok(new { status = "success", message = "update successfully!", user = result });

            }
            catch (ArgumentException e)
            {
                // Bắt lỗi ArgumentException và trả về phản hồi BadRequest với thông điệp lỗi
                return BadRequest(new { status = "error", message = e.Message });
            }
            catch (Exception e)
            {
                return StatusCode(500, new { status = "error", message = "An error occurred while processing your request", error = e.Message });
            }
        }


        [HttpPost]
        [Route("reset-password")]

        public async Task<IActionResult> ResetPassword([FromForm] int user_id, string oldPassword, string newPassword)
        {
            try
            {
                var result = await _authService.reset_password(user_id, oldPassword, newPassword);

                if (!result)
                {
                    return BadRequest(new { status = "error", message = "Thay đổi mật khẩu thất bại." });
                }

                return Ok(new { status = "success", message = "Reset successfully!" });

            }
            catch (ArgumentException e)
            {
                // Bắt lỗi ArgumentException và trả về phản hồi BadRequest với thông điệp lỗi
                return BadRequest(new { status = "error", message = e.Message });
            }
            catch (Exception e)
            {
                return StatusCode(500, new { status = "error", message = "An error occurred while processing your request", error = e.Message });
            }
        }



    }
}

File: font_end/fruitsapp_admin/src/Controllers/authController.js

import AuthRequest from "API/AuthRequest/authRequest";
import User from "Models/User";
import ErrorHandler from "Utils/errorHandler";


const AuthController = {
    login: async (login_infor) => {
        try {
            const response = await AuthRequest.login(login_infor);
            const result = response.data;
            if (result.status === 'success') {
                return { success: true, token: result.token, message: result.message };
            } else {
                return { success: false, message: result.message };
            }
        } catch (error) {
            return ErrorHandler.handle(error);
        }
    },
    register: async (formData) => {
        try {
            const response = await AuthRequest.register(formData);
            const result = response.data;
            if (result.status === 'success') {
                return { success: true, data: result.data, message: result.message };
            } else {
                return { success: false, message: result.message };
            }
        } catch (error) {
            return ErrorHandler.handle(error);
        }
    },
    getProfileUser: async () => {
        try {
            const response = await AuthRequest.get_user();
            const result = response.data;
            if (result.status === 'success') {
                const userData = new User(
                    result.user.id,
                    result.user.role_id,
                    result.user.name,
                    result.user.avatar_url,
                    result.user.email,
                    result.user.phone,
                    result.user.create_at,
                    result.user.update_at,
                    result.user.disable,
                    result.user.role);
                return { success: true, user: userData, message: result.message };
            } else {
                return { success: false, message: result.message };
            }
        } catch (error) {
            return ErrorHandler.handle(error);
        }
    },
    logout: async(token)=>{
        try{
            const response = await AuthRequest.logout(token);
            const result = response.data;
            if(result.status === 'success'){
                return { success: true, message: result.message };
            }else{
                return { success: false, message: result.message };
            }

        }catch(error){
            return ErrorHandler.handle(error);
        }
    },
    update: async(formData)=>{
        try{
            const response = await AuthRequest.update(formData);
            const result = response.data;
            if(result.status === 'success'){
                const userData = new User(
                    result.user.id,
                    result.user.role_id,
                    result.user.name,
                    result.user.avatar_url,
                    result.user.email,
                    result.user.phone,
                    result.user.create_at,
                    result.user.update_at,
                    result.user.disable,
                    result.user.role);
                return { success: true, user: userData ,message: result.message };
            }else{
                return { success: false, message: result.message };
            }

        }catch(error){
            return ErrorHandler.handle(error);
        }
    },
    ChangePass: async(formData)=>{
        try{
            const response = await AuthRequest.change_password(formData);
            const result = response.data;
            if(result.status === 'success'){
                
                return { success: true ,message: result.message };
            }else{
                return { success: false, message: result.message };
            }

        }catch(error){
            return ErrorHandler.handle(error);
        }
    }
}

export default AuthController;